src/ctap/clock.rs

```diff
@@ -0,0 +1,55 @@
+pub use embedded_time::Clock;
+use libtock_drivers::result::FlexUnwrap;
+use libtock_drivers::timer::Timer;
+
+pub struct LibtockClock<'a, const CLOCK_FREQUENCY: u32>(Timer<'a>);
+impl<'a, const CLOCK_FREQUENCY: u32> LibtockClock<'a, CLOCK_FREQUENCY> {
+    pub fn new(timer: Timer<'a>) -> Self {
+        Self(timer)
+    }
+}
+
+#[cfg(target_pointer_width = "32")]
+pub type ClockInt = u32;
+#[cfg(target_pointer_width = "64")]
+pub type ClockInt = u64;
+
+impl<const CLOCK_FREQUENCY: u32> embedded_time::Clock for LibtockClock<'_, CLOCK_FREQUENCY> {
+    // TODO: Implement and use a 24-bits TimeInt for Nordic
+    type T = ClockInt;
+
+    const SCALING_FACTOR: embedded_time::fraction::Fraction =
+        <embedded_time::fraction::Fraction>::new(1, CLOCK_FREQUENCY);
+
+    fn try_now(&self) -> Result<embedded_time::Instant<Self>, embedded_time::clock::Error> {
+        let timer = &self.0;
+        let now = timer.get_current_clock().flex_unwrap();
+        Ok(embedded_time::Instant::new(now.num_ticks() as Self::T))
+    }
+}
+
+pub type CtapClock<'a> = LibtockClock<'a, 32768>;
+
+pub fn new_clock<'a>(timer: Timer<'a>) -> CtapClock {
+    CtapClock::new(timer)
+}
+
+//pub type CtapInstant<'a> = embedded_time::Instant<CtapClock<'a>>;
```

Comment:
Mmm, I fear that this lifetime parameter of `Timer` might be problematic in the future. You might need to make some code parametric over this lifetime. If you do, please ping me with the locations. I'd like to take a look to see if there's a way to work around or factorize this effort for other hardware components.

Reply:
Here's a solution to remove the lifetime: https://github.com/ia0/OpenSK/commit/eb220320d4a21bb12a2df1fb21329ee4bb78e690. I think it's ok to go with this until the code is generic. At which point it might be easier to choose a different solution.

Reply:
If we plan to replace it eventually, that might be okay. Otherwise, is it necessary to introduce unsafe code?

Reply:
Yes, any solution to hide the fact that the libtock clock needs a lifetime (which is just a consequence of its design asking a useless callback) is essentially unsafe somewhere. The 2 obvious solutions are:
1. Using static lifetime. This solution. I think the best solution given its simplicity.
2. Using `Pin`. The unsafe would be hidden under the hood though. But we could also hide the `unsafe` keyword under the hood for option 1 too. It's actually what we do for the storage:
    - No `unsafe` keyword but actually unsafe (I managed to get the English wrong in a single sentence, good job): https://github.com/google/OpenSK/blob/d47ca7fa5437e879b25f5b77e17f83118c4f1b51/src/ctap/storage.rs#L68
    - The actual `unsafe` keyword is much deeper: https://github.com/google/OpenSK/blob/d47ca7fa5437e879b25f5b77e17f83118c4f1b51/src/embedded_flash/syscall.rs#L164

Long term, all those "call only once" will disappear because the library won't create the hardware. The hardware will be created by the user and passed to the library. In embedded world it's pretty typical to have this "use once" constraint. The way it's done in embedded Rust is through a [`Peripherals::take`](https://docs.rs/nrf52840-hal/latest/nrf52840_hal/pac/struct.Peripherals.html#method.take) which only returns `Some` the first time and then `None`. We could also do that here if we want to get rid of the surfacing `unsafe` keyword. But this is only temporary until we have a library, so I'm not sure it's worth it.

Reply:
Are we allowed to use the alloc crate? If so, we can leak instead of being unsafe:
```rust
pub fn new() -> Self {
       let boxed_cb = alloc::boxed::Box::new(timer::with_callback(|_, _| {}));
       let timer = alloc::boxed::Box::leak(boxed_cb).init().flex_unwrap();
       Self(timer)
}
```
Assuming we call it once, the leak doesn't matter, and if we call it more than once, at least we do not trash the memory. We could also track the number of time it is called and panic if it is called twice.

Reply:
Yes this is a safe option, sounds good. Note we could also use the [`once`](https://docs.rs/once/latest/once/macro.assert_has_not_been_called.html) crate if we want to check that it's called once.